metricbeat/module/prometheus/collector/collector.go

```diff
@@ -186,6 +194,16 @@ func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
 
 // Close stops the metricset
 func (m *MetricSet) Close() error {
+	attempts := 0
+	for !m.eventGenStarted {
+		time.Sleep(m.period)
+		// waiting 2 periods so as the first Fetch to start the generator, otherwise return since
+		// Fetch is not called at all
+		if attempts > 2 {
+			return nil
+		}
+		attempts++
+	}
```

Comment:
I don't think you need a backoff here. MB will make sure Fetch is not called after Close. This could just be:

if !m.eventGenStarted {
   m.promEventsGen.Stop()
}